// stylelint/dist/rules/at-rule-no-unknown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (actual, options) {
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, { actual: actual }, {
      actual: options,
      possible: {
        ignoreAtRules: [_lodash.isString]
      },
      optional: true
    });

    if (!validOptions) {
      return;
    }

    root.walkAtRules(function (atRule) {
      var nameAtRule = atRule.name;

      if (_postcss.vendor.prefix(nameAtRule) || _keywordSets.atRules.has(nameAtRule.toLowerCase())) {
        return;
      }

      var ignoredAtRules = options && options.ignoreAtRules || [];

      if (ignoredAtRules.indexOf(nameAtRule.toLowerCase()) !== -1) {
        return;
      }

      (0, _utils.report)({
        message: messages.rejected("@" + nameAtRule),
        node: atRule,
        ruleName: ruleName,
        result: result
      });
    });
  };
};

var _lodash = require("lodash");

var _postcss = require("postcss");

var _utils = require("../../utils");

var _keywordSets = require("../../reference/keywordSets");

var ruleName = exports.ruleName = "at-rule-no-unknown";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(atRule) {
    return "Unexpected unknown at-rule \"" + atRule + "\"";
  }
});
