// stylelint/dist/rules/selector-attribute-operator-whitelist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.messages = exports.ruleName = undefined;

exports.default = function (whitelistInput) {
  var whitelist = [].concat(whitelistInput);
  return function (root, result) {
    var validOptions = (0, _utils.validateOptions)(result, ruleName, {
      actual: whitelist,
      possible: [_lodash.isString]
    });
    if (!validOptions) {
      return;
    }

    root.walkRules(function (rule) {
      if (!(0, _utils.isStandardSyntaxRule)(rule)) {
        return;
      }
      if (rule.selector.indexOf("[") === -1 || rule.selector.indexOf("=") === -1) {
        return;
      }

      (0, _utils.parseSelector)(rule.selector, result, rule, function (selectorTree) {
        selectorTree.walkAttributes(function (attributeNode) {
          var operator = attributeNode.operator;

          if (!operator || operator && whitelist.indexOf(operator) !== -1) {
            return;
          }

          (0, _utils.report)({
            message: messages.rejected(operator),
            node: rule,
            index: attributeNode.attribute.length + 1,
            result: result,
            ruleName: ruleName
          });
        });
      });
    });
  };
};

var _lodash = require("lodash");

var _utils = require("../../utils");

var ruleName = exports.ruleName = "selector-attribute-operator-whitelist";

var messages = exports.messages = (0, _utils.ruleMessages)(ruleName, {
  rejected: function rejected(operator) {
    return "Unexpected operator \"" + operator + "\"";
  }
});
